File: day1/day1.go
package main

import (
	"fmt"
	"math"
	"os"
	"slices"
	"strconv"
	"strings"
)

func day1(input []byte) {
	s := string(input)
	lines := strings.Split(s, "\n")
	leftNums := []int{}
	rightNums := []int{}
	sum := 0

	for _, line := range lines {
		lineVals := len((strings.Split(line, " ")))

		leftNum, _ := strconv.Atoi(strings.Split(line, " ")[0])
		leftNums = append(leftNums, leftNum)

		rightNum, _ := strconv.Atoi(strings.Split(line, " ")[lineVals-1])
		rightNums = append(rightNums, rightNum)
	}

	slices.Sort(leftNums)
	slices.Sort(rightNums)

	for i, leftNum := range leftNums {
		rightNum := rightNums[i]
		dist := int(math.Abs(float64(leftNum - rightNum)))
		sum += dist
	}
	fmt.Printf("The result for day 1 is: %d", sum)
}

func day2(input []byte) {
	s := string(input)
	lines := strings.Split(s, "\n")
	simScore := 0
	rightOccurs := make(map[int]int)

	leftNums := []int{}

	for _, line := range lines {
		lineVals := len((strings.Split(line, " ")))

		leftNum, _ := strconv.Atoi(strings.Split(line, " ")[0])
		leftNums = append(leftNums, leftNum)

		rightNum, _ := strconv.Atoi(strings.Split(line, " ")[lineVals-1])
		rightOccurs[rightNum] += 1
	}

	for _, num := range leftNums {
		simScore += (num * rightOccurs[num])
	}

	fmt.Printf("The result for day 2 is: %d", simScore)
}

func main() {
	input, err := os.ReadFile("input.txt")
	if err != nil {
		return
	}
	day1(input)
	day2(input)
}

File: day1/day1.py
import os

def part1(lines):
    sum = 0
    rnums = []
    lnums = []
    for line in lines:
        nums = line.split("   ")
        lnum = int(nums[0])
        rnum = int(nums[1])
        lnums.append(lnum)
        rnums.append(rnum)

    rnums.sort()
    lnums.sort()
    
    for i, lnum in enumerate(lnums):
        sum += abs(lnum - rnums[i])
    print(sum)

def part2(lines):
    sum = 0
    rnumFreq = {}

    for line in lines:
        rnum = int(line.split("   ")[1])
        rnumFreq[rnum] = rnumFreq.get(rnum, 0) + 1
    
    for line in lines:
        nums = line.split("   ")
        lnum = int(nums[0])
        rnum = int(nums[1]) 
        if lnum in rnumFreq:
            sum += lnum * rnumFreq[lnum]

    print(sum)

with open("input.txt", 'r') as inFile:
    lines = inFile.read().split('\n')
    part1(lines)
    part2(lines)

